### ISSUE TITLE
odd selection of points in demo

### ISSUE THREAD

**Issue by kkdd** [flekschas/regl-scatterplot#5]
Hello,
I have noticed that some odd points are highlighted, even though they are not selected by a mouse click/lasso in the [demo](https://flekschas.github.io/regl-scatterplot/):

![scatter](https://user-images.githubusercontent.com/5372642/71613246-9cb90680-2be8-11ea-9864-a9ba5dfa1570.jpg)

I'm using Google Chrome 79.0.3945.88.

**Comment by flekschas** [comment]
Thanks for reporting! I thought I fixed that issue some time ago but apparently I did not. The issue is related to the texture state.

What exactly did you do to get to this spread situation? Did you lasso select several dots or just click on one?

**Comment by kkdd** [comment]
Thank you for your reply.
I lasso-selected dots in a small area in the example reported.
(I struggled but I failed to capture a screen shot with a mouse cursor/loci.)

**Comment by flekschas** [comment]
Thanks. I'll take a look

**Comment by alejandrokennedy** [comment]
Hi @flekschas and @kkdd! I am experiencing the same problem: whether hovering, clicking, or lassoing, it seems like the correct point is selected/highlighted only less than 10% of the time. Here is a [gDrive link to a video](https://drive.google.com/open?id=1uljz96YBEug-GJ7wAs4uDN-4iSG9TF4U) that demonstrates this behaviour. To reproduce it, I simply pulled the project and ran npm start to view the example. I didn't change any code. It also happens with lassoing on the [online example](https://flekschas.github.io/regl-scatterplot/).

I am happy to help troubleshoot, as I was working on my own implementation of a regl scatterplot when I found this project! I am new to regl, though, so I'm limited by that. Still, I prefer the idea of contributing to this project to starting my own. Please let me know what I can do to help!

**Comment by flekschas** [comment]
@alejandrokennedy Thanks for chiming in! And sorry I never looked into this. I know that the general issue is related to the floating-point precision as I am using a float-based texture as the point state model. I thought I had fixed that issue earlier but it seems I was wrong.

The related code is this: https://github.com/flekschas/regl-scatterplot/blob/master/src/point.vs#L49-L52

Essentially each pixel in that texture represents a point and the value of that pixel represents the point state (i.e., normal, hover, active, etc.)

@alejandrokennedy I am happy for any help. 

I'll try to look into this over the weekend.

**Comment by alejandrokennedy** [comment]
@flekschas, thanks for the quick reply! If it's any consolation, I just ran npm start on my work machine. Same exact code (synced by dropbox), but on my work computer the selection behaviour is perfect—no errant selections. I'm not sure what the difference could be, but I'm happy to help troubleshoot if you're having trouble reproducing the error.

**Comment by flekschas** [comment]
Thanks for testing! That might actually tell us that the issue is related to some WebGL / graphics card setting rather than the code itself.

**Comment by flekschas** [comment]
I was able to reproduce the issue when I bumped up the number of points to 1 million.

My best guess is that the floating-point precision is too low for some reason. When I change the following line https://github.com/flekschas/regl-scatterplot/blob/master/src/point.vs#L2 to 

```glsl
precision highp float;
```

I am not able to reproduce the error anymore, even with 1 million points.

@kkdd and @alejandrokennedy does that solve the issue for you guys as well?

**Comment by alejandrokennedy** [comment]
It doesn't seem to have solved the issue, but I'm not entirely sure I'm testing correctly (I don't have much experience with pull requests). I followed [this page](https://github.com/genome/docs/wiki/How-do-I-%60git-checkout%60-a-pull-request%3F) for instructions, and am attaching a screenshot of my terminal:

![Screen Shot 2020-01-31 at 10 56 26 AM](https://user-images.githubusercontent.com/6080618/73553697-5d185100-4418-11ea-9439-afa95c326a77.png)

...so that hopefully you can let me know if I was testing correctly. Once I checked out the pull request, I ran npm start, which loaded the example page. On my older machine (MacBook Pro (Retina, 15-inch, Mid 2014), 2.5 GHz Intel Core i7, 16 GB 1600 MHz DDR3, NVIDIA GeForce GT 750M 2 GB Intel Iris Pro 1536 MB), I experience the same glitchy behaviour you can see in the video from my first comment. On my newer machine, however (MacBook Pro (13-inch, 2017, Two Thunderbolt 3 ports), 2.3 GHz Intel Core i5, 16 GB 2133 MHz LPDDR3, Intel Iris Plus Graphics 640 1536 MB), I don't see that behaviour. Interestingly, it doesn't matter whether I use chrome, firefox, or opera. The machine seems to be the deciding factor (all browsers on the old machine fail to select appropriate points, and all browsers on the new machine do it correctly). Both are running Mojave.

Thoughts?

**Comment by flekschas** [comment]
What you're describing supports my hypothesis that it's some kind of GL/graphics card setting that is causing the lasso selection to be thrown off.

I am not sure what branch you're on from the screenshot. You should be able to do the following for testing:

```bash
git clone https://github.com/flekschas/regl-scatterplot/
cd regl-scatterplot
git fetch
git checkout increase-floating-point-precision
npm i
npm run start
```

You can verify if you're in the correct branch by looking at the second line in `src/point.vs`. It should look as follows:

```glsl
precision highp float;
```

Alternatively you can also just change that line to the one above in your current branch.

**Comment by alejandrokennedy** [comment]
Thanks for the info! I'll double check tomorrow that I was on the right branch. Let me know if there are any tests I can perform besides simply observing glitchy selections.

**Comment by flekschas** [comment]
Thanks! There aren't any _hard_ tests we can before because the issue is only visually (the state itself is correct, I've tested that already).

Just to describe the issue a bit better: imagine we have 4 dots. To keep track of the state (i.e., position, state, and value of the dots) regl-scatterplot is created an RGBA float texture. We always create a square texture, so in this example, we will have a 2x2x4 array. The length of either side of the texture is what we call its _resolution_ (or _res_ for short). We'll need that later. The 4 values of the RGBA color then encode the following values (in this given order): x coordinate, y-coordinate, color state, a numerical value. The x and y coordinate should be clear. Now regl-scatterplot supports two distinct coloring modes: categorical coloring and numerical coloring. For categorical color, i.e., when we only want to use a small number of distinct colors, we will use the color state, where the states correspond with our colors. The color state is an thus integer (represented as a float) indicating what color a dot should have.

Note the actual RGBA value of the color is stored in another texture objects. This allows us to save some memory. E.g., imagine we gave 1M dots but only 2 classes. Using the color states we only need 4 * 1M floats (for the base state texture) plus 2 * 4 floats for the 2 colors instead of (2 + 4) * 1M floats for the xy position and individual RGBA values for each dot. So we saved 3 * 1M floats.). For numerical coloring we'll do something very similar by taking the numerical value of a dot and looking up its color in a colormap texture.

To come back to the actual problem. In order to determine in which color we want to draw points, we'll have to get the correct state, which means we have to select the correct pixel of the state texture object. To do so we have to convert the index of the dot (which we pass into the shader) and determine the corresponding i,j position of the pixel. Given our toy example from above, if we wanted to draw the second dot this would look as follows:

```glsl
uniform float stateTexRes;
// -> This is the width/height of the state texture, i.e., 2 in our example

attribute float stateIndex;
// -> This is 1 as we're drawing the second dot

void main() {
  float eps = 0.5 / stateTexRes; // Have a texel in texture coordinates
  // -> 0.25

  // Remember the texture shape is a squareish as possible so it's 2x2. So we need to get the row and column of the pixel corresponding to our dot
  float stateRowIndex = floor((stateIndex + eps) / stateTexRes);
  // -> floor((1 + 0.25) / 2) = floor(.625) = 0
  // The second pixel should be in the first row!

  // Now we construct a tuple for accessing the state texture. Note that this tuple is in texture coordinates
  vec2 stateTexIndex = vec2(
    (stateIndex / stateTexRes) - stateRowIndex + eps,
    stateRowIndex / stateTexRes + eps
  );
  // -> [(1 / 2) - 0 + 0.25, 0 / 2] = [0.75, 0.25]
  // This represents the center of the top-right pixel!
}
```

Now, somehow it can happen that we're accessing the incorrect pixel! If that happens we're highlighting the dots incorrectly! You might wonder why we need `eps`? If we don't do this, strange things can happen due to limited floating-point precision. E.g., `[(1 / 2) - 0, 0 / 2]` == [0.5, 0]` but if `[(0.999 / 2.001) - 0, 0 / 2] = [0.499, 0]`, which corresponds to the top-left and not the top-right pixel!

I thought that the addition of `eps` solved the issue but it seems like it did not.
